Kernel/Storage/NVMe/NVMeQueue.cpp

```diff
@@ -114,24 +114,12 @@ u16 NVMeQueue::submit_sync_sqe(NVMeSubmission& sub)
 
         if (m_requests.contains(sub.cmdid) && m_requests.get(sub.cmdid).release_value().used)
             VERIFY_NOT_REACHED();
-        m_requests.set(sub.cmdid, { nullptr, true, nullptr });
+        m_requests.set(sub.cmdid, { nullptr, true, [this, &cmd_status](u16 status) mutable { cmd_status = status; m_sync_wait_queue.wake_all(); } });
     }
-
     submit_sqe(sub);
-    do {
-        int index;
-        {
-            SpinlockLocker lock(m_cq_lock);
-            index = m_cq_head - 1;
-            if (index < 0)
-                index = m_qdepth - 1;
-        }
-        cqe_cid = m_cqe_array[index].command_id;
-        microseconds_delay(1);
-    } while (cid != cqe_cid);
 
-    auto status = CQ_STATUS_FIELD(m_cqe_array[m_cq_head].status);
-    return status;
+    m_sync_wait_queue.wait_forever("NVMe sync submit"sv);
```

Comment:
I am not fond of this pattern because this inherently says that we could be technically stuck forever for this thread. I know this has nothing to do with your PR (this is a poor design we need to fix), just saying this here loud and clear.

Reply:
I completely agree with you. But note that I do this only for submit_sync_sqe which is used only by the admin queue.

Normal IO path goes via submit_sqe which doesn't have a wait queue based IO ( at least in the driver code. I know block layer follows this pattern for Normal IO).

Eventually we need to move all this logic to the block layer instead of sprinkling it in driver code.

Reply:
Can we capture this idea as a FIXME in a comment next to the relevant code?